File: frontend/js/offers.js
/**
 * Offers: which ones a box has earned, and what each is worth at checkout.
 *
 * Every offer in the catalog carries a `rule`; this module is the only place
 * that reads it. Nothing here is a blanket coupon: an offer is either eligible
 * against the current box, with the amount it takes off, or ineligible with
 * the plain reason why. One discount applies at a time. Free delivery is
 * automatic and stacks with whichever discount is applied.
 *
 * Pure: takes the box and the customer, returns numbers. The server evaluates
 * the same rules in backend/offers.py.
 */

const money = (value) => `$${value.toFixed(2)}`;
const round = (value) => Math.round(value * 100) / 100;

/** Facts every rule needs, computed once from the box. */
function facts(basket, context) {
  const lines = basket.lines || [];
  const count = basket.count ?? lines.reduce((sum, line) => sum + (line.qty || 1), 0);
  const subtotal = basket.subtotal ?? lines.reduce((sum, line) => sum + line.item.price * (line.qty || 1), 0);
  const liveSeasons = context.liveSeasons || [];
  const liveIds = new Set(liveSeasons.map((season) => season.id));
  const seasonalUnits = lines
    .filter((line) => line.item.season && (liveIds.size === 0 || liveIds.has(line.item.season)))
    .reduce((sum, line) => sum + (line.qty || 1), 0);
  const favorites = context.favorites || [];
  const favoriteIds = new Set(favorites.map((f) => f.id));
  const favoriteLines = lines
    .filter((line) => favoriteIds.has(line.item.id))
    .sort((a, b) => (b.qty || 1) - (a.qty || 1));
  return {
    lines, count, subtotal,
    flavors: new Set(lines.map((line) => line.item.id)).size,
    seasonalUnits,
    liveSeasonName: liveSeasons[0]?.name || null,
    signedIn: Boolean(context.signedIn),
    repeatPattern: context.repeatPattern || null,
    favorites,
    favoriteLines,
  };
}

const plural = (n, word) => `${n} ${word}${n === 1 ? "" : "s"}`;

/** Evaluate one rule. Returns { eligible, discount, why, progress, addon }. */
function evaluate(offer, f) {
  const rule = offer.rule || {};
  switch (rule.type) {
    case "percent": {
      if (rule.requires === "favorite") {
        if (!f.signedIn) return { eligible: false, why: "Sign in and your reorder rate applies to a box of favorites." };
        if (!f.favorites.length) return { eligible: false, why: "Order a couple of boxes first — this one prices your regulars." };
        const need = rule.minFavoriteUnits || 4;
        const hit = f.favoriteLines.find((line) => (line.qty || 1) >= need);
        if (hit) {
          return {
            eligible: true, discount: round(f.subtotal * rule.percent / 100),
            why: `${hit.qty} × ${hit.item.name} — you keep coming back for it.`,
          };
        }
        const closest = f.favoriteLines[0];
        const favoriteName = closest?.item.name || f.favorites[0].name || "a favorite";
        const have = closest?.qty || 0;
        return {
          eligible: false,
          why: `Add ${plural(need - have, "more")} ${favoriteName} — ${need} of a favorite unlocks it.`,
          progress: Math.round((have / need) * 100),
        };
      }
      const tiers = rule.tiers || [{ minItems: rule.minItems || 0, percent: rule.percent }];
      const tier = tiers.find((t) => f.count >= (t.minItems || 0) && f.flavors >= (t.minFlavors || 0));
      if (tier) {
        const better = tiers.find((t) => t.percent > tier.percent);
        const nudge = better && f.flavors < (better.minFlavors || 0)
          ? ` Mix in ${plural((better.minFlavors || 0) - f.flavors, "more flavor")} for ${better.percent}%.`
          : "";
        return {
          eligible: true, discount: round(f.subtotal * tier.percent / 100), percent: tier.percent,
          why: `${f.count} treats across ${plural(f.flavors, "flavor")} — the box takes ${tier.percent}% off.${nudge}`,
        };
      }
      const min = Math.min(...tiers.map((t) => t.minItems || 0));
      return {
        eligible: false,
        why: `Add ${plural(min - f.count, "more")} to fill the six-count.`,
        progress: Math.round((f.count / min) * 100),
      };
    }
    case "free-addon": {
      const min = rule.minItems || 12;
      if (f.count >= min) {
        return { eligible: true, discount: 0, addon: rule.addon, why: `${f.count} treats — the ${rule.addon.name.toLowerCase()} rides along free.` };
      }
      return { eligible: false, why: `Add ${plural(min - f.count, "more")} for the party box.`, progress: Math.round((f.count / min) * 100) };
    }
    case "fixed-per-item": {
      if (rule.requires === "seasonal") {
        if (!f.liveSeasonName) return { eligible: false, why: "No seasonal menu is live right now." };
        if (f.seasonalUnits === 0) return { eligible: false, why: `Add something from ${f.liveSeasonName} to use it.` };
        const units = Math.min(f.seasonalUnits, rule.maxUnits || Infinity);
        return {
          eligible: true, discount: round(rule.amount * units),
          why: `${plural(units, "seasonal treat")} in the box at ${money(rule.amount)} off each.`,
        };
      }
      return { eligible: false, why: "Not available." };
    }
    case "free-delivery": {
      const min = rule.minSubtotal || 45;
      if (f.subtotal >= min) {
        return { eligible: true, discount: 0, why: "Applied on its own when you choose delivery at checkout.", progress: 100 };
      }
      return {
        eligible: false,
        why: `${money(min - f.subtotal)} more and delivery is free. Applies on its own at checkout.`,
        progress: Math.round((f.subtotal / min) * 100),
      };
    }
    default:
      return { eligible: false, why: "Not available." };
  }
}

/**
 * Every offer, with whether the box has earned it and what it is worth.
 * Eligible ones come first, best value first.
 */
export function evaluateOffers(offers, basket, context = {}) {
  const f = facts(basket, context);
  return offers
    .map((offer) => {
      const result = evaluate(offer, f);
      return {
        ...offer,
        eligible: result.eligible,
        live: result.eligible,
        discount: result.discount || 0,
        addon: result.addon || null,
        why: result.why,
        progress: result.progress ?? (result.eligible ? 100 : null),
        percent: result.percent,
      };
    })
    .sort((a, b) => Number(b.eligible) - Number(a.eligible) || b.discount - a.discount);
}

/** The discount the box would take if the customer applied the best one. */
export function bestOffer(evaluated) {
  return evaluated.find((offer) => offer.eligible && !offer.auto && offer.discount > 0) || null;
}

/**
 * The receipt. `appliedId` is honored only while that offer is still
 * eligible; the caller learns from `applied` whether it survived.
 */
export function priceBox(offers, basket, { appliedId = null, fulfillment = "pickup", context = {} } = {}) {
  const evaluated = evaluateOffers(offers, basket, context);
  const f = facts(basket, context);
  const applied = appliedId ? evaluated.find((offer) => offer.id === appliedId && offer.eligible && !offer.auto) || null : null;
  const delivery = offers.find((offer) => offer.rule?.type === "free-delivery");
  const fee = delivery?.rule?.fee ?? 6.95;
  const threshold = delivery?.rule?.minSubtotal ?? 45;
  const deliveryFree = f.subtotal >= threshold;
  const deliveryFee = fulfillment === "delivery" && !deliveryFree ? fee : 0;
  const discount = applied ? applied.discount : 0;

  return {
    subtotal: round(f.subtotal),
    applied,
    discount: round(discount),
    addon: applied?.addon || null,
    fulfillment,
    deliveryFee: round(deliveryFee),
    deliveryFree: fulfillment === "delivery" && deliveryFree,
    total: round(Math.max(0, f.subtotal - discount + deliveryFee)),
    evaluated,
  };
}

File: backend/offers.py
"""Offer rules, evaluated the way the browser does (frontend/js/offers.js).

Every offer in the catalog carries a `rule`. Given the box and what is known
about the customer, each offer is either eligible with the amount it takes
off, or ineligible with the plain reason why. One discount applies at a time;
free delivery is automatic and stacks.

Lines are [{"item": <catalog item>, "qty": int}].
"""


def _money(value):
    return f"${value:.2f}"


def _plural(n, word):
    return f"{n} {word}{'' if n == 1 else 's'}"


def _facts(lines, customer, live_seasons):
    customer = customer or {}
    count = sum(l["qty"] for l in lines)
    subtotal = sum(l["item"]["price"] * l["qty"] for l in lines)
    live_ids = {s["id"] for s in live_seasons}
    seasonal_units = sum(
        l["qty"] for l in lines
        if l["item"].get("season") and (not live_ids or l["item"]["season"] in live_ids))
    favorites = customer.get("favorites") or []
    favorite_ids = {f["id"] for f in favorites}
    favorite_lines = sorted(
        (l for l in lines if l["item"]["id"] in favorite_ids),
        key=lambda l: -l["qty"])
    return {
        "lines": lines, "count": count, "subtotal": subtotal,
        "flavors": len({l["item"]["id"] for l in lines}),
        "seasonal_units": seasonal_units,
        "live_season_name": live_seasons[0]["name"] if live_seasons else None,
        "signed_in": bool(customer.get("signed_in")),
        "favorites": favorites, "favorite_lines": favorite_lines,
    }


def _evaluate(offer, f):
    rule = offer.get("rule") or {}
    kind = rule.get("type")

    if kind == "percent":
        if rule.get("requires") == "favorite":
            if not f["signed_in"]:
                return {"eligible": False, "why": "Sign in and your reorder rate applies to a box of favorites."}
            if not f["favorites"]:
                return {"eligible": False, "why": "Order a couple of boxes first — this one prices your regulars."}
            need = rule.get("minFavoriteUnits", 4)
            hit = next((l for l in f["favorite_lines"] if l["qty"] >= need), None)
            if hit:
                return {"eligible": True, "discount": round(f["subtotal"] * rule["percent"] / 100, 2),
                        "why": f"{hit['qty']} × {hit['item']['name']} — you keep coming back for it."}
            closest = f["favorite_lines"][0] if f["favorite_lines"] else None
            name = closest["item"]["name"] if closest else f["favorites"][0].get("name", "a favorite")
            have = closest["qty"] if closest else 0
            return {"eligible": False, "progress": round(have / need * 100),
                    "why": f"Add {_plural(need - have, 'more')} {name} — {need} of a favorite unlocks it."}

        tiers = rule.get("tiers") or [{"minItems": rule.get("minItems", 0), "percent": rule["percent"]}]
        tier = next((t for t in tiers
                     if f["count"] >= t.get("minItems", 0) and f["flavors"] >= t.get("minFlavors", 0)), None)
        if tier:
            better = next((t for t in tiers if t["percent"] > tier["percent"]), None)
            nudge = ""
            if better and f["flavors"] < better.get("minFlavors", 0):
                nudge = f" Mix in {_plural(better['minFlavors'] - f['flavors'], 'more flavor')} for {better['percent']}%."
            return {"eligible": True, "percent": tier["percent"],
                    "discount": round(f["subtotal"] * tier["percent"] / 100, 2),
                    "why": f"{f['count']} treats across {_plural(f['flavors'], 'flavor')} — the box takes {tier['percent']}% off.{nudge}"}
        need = min(t.get("minItems", 0) for t in tiers)
        return {"eligible": False, "progress": round(f["count"] / need * 100) if need else 0,
                "why": f"Add {_plural(need - f['count'], 'more')} to fill the six-count."}

    if kind == "free-addon":
        need = rule.get("minItems", 12)
        if f["count"] >= need:
            return {"eligible": True, "discount": 0, "addon": rule.get("addon"),
                    "why": f"{f['count']} treats — the {rule['addon']['name'].lower()} rides along free."}
        return {"eligible": False, "progress": round(f["count"] / need * 100),
                "why": f"Add {_plural(need - f['count'], 'more')} for the party box."}

    if kind == "fixed-per-item" and rule.get("requires") == "seasonal":
        if not f["live_season_name"]:
            return {"eligible": False, "why": "No seasonal menu is live right now."}
        if f["seasonal_units"] == 0:
            return {"eligible": False, "why": f"Add something from {f['live_season_name']} to use it."}
        units = min(f["seasonal_units"], rule.get("maxUnits", f["seasonal_units"]))
        return {"eligible": True, "discount": round(rule["amount"] * units, 2),
                "why": f"{_plural(units, 'seasonal treat')} in the box at {_money(rule['amount'])} off each."}

    if kind == "free-delivery":
        need = rule.get("minSubtotal", 45)
        if f["subtotal"] >= need:
            return {"eligible": True, "discount": 0, "progress": 100,
                    "why": "Applied on its own when you choose delivery at checkout."}
        return {"eligible": False, "progress": round(f["subtotal"] / need * 100),
                "why": f"{_money(need - f['subtotal'])} more and delivery is free. Applies on its own at checkout."}

    return {"eligible": False, "why": "Not available."}


def evaluate_offers(offers, lines, customer=None, live_seasons=()):
    """Every offer with eligibility, value and reason. Eligible first."""
    f = _facts(lines, customer, list(live_seasons))
    out = []
    for offer in offers:
        result = _evaluate(offer, f)
        out.append({
            **offer,
            "eligible": result["eligible"],
            "discount": result.get("discount", 0),
            "addon": result.get("addon"),
            "why": result["why"],
            "progress": result.get("progress", 100 if result["eligible"] else None),
        })
    out.sort(key=lambda o: (not o["eligible"], -o["discount"]))
    return out


def price(offers, lines, applied_id=None, fulfillment="pickup", customer=None, live_seasons=()):
    """The receipt: subtotal, the applied discount if it still holds, delivery, total."""
    evaluated = evaluate_offers(offers, lines, customer, live_seasons)
    f = _facts(lines, customer, list(live_seasons))
    applied = next((o for o in evaluated
                    if o["id"] == applied_id and o["eligible"] and not o.get("auto")), None)
    delivery = next((o for o in offers if (o.get("rule") or {}).get("type") == "free-delivery"), None)
    fee = (delivery or {}).get("rule", {}).get("fee", 6.95)
    threshold = (delivery or {}).get("rule", {}).get("minSubtotal", 45)
    delivery_free = f["subtotal"] >= threshold
    delivery_fee = fee if fulfillment == "delivery" and not delivery_free else 0
    discount = applied["discount"] if applied else 0
    return {
        "subtotal": round(f["subtotal"], 2),
        "applied": applied,
        "discount": round(discount, 2),
        "addon": applied.get("addon") if applied else None,
        "fulfillment": fulfillment,
        "deliveryFee": round(delivery_fee, 2),
        "deliveryFree": fulfillment == "delivery" and delivery_free,
        "total": round(max(0, f["subtotal"] - discount + delivery_fee), 2),
    }
